src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoverer.cs

```diff
@@ -140,9 +157,138 @@ private List<PluginFile> GetPluginFiles(CancellationToken cancellationToken)
             return files;
         }
 
+        /// <summary>
+        /// Retrieves authentication plugins by searching through directories specified in the `NuGET_PLUGIN_PATHS` and `PATH` 
+        /// environment variables. The method looks for files prefixed with 'nuget-plugin-' and verifies their validity for .net tools plugins.
+        /// </summary>
+        /// <returns>A list of valid <see cref="PluginFile"/> objects representing the discovered plugins.</returns>
+        internal List<PluginFile> GetPluginsInNuGetPluginPathsAndPath()
+        {
+            var pluginFiles = new List<PluginFile>();
+            var nugetPluginPaths = _environmentVariableReader.GetEnvironmentVariable(EnvironmentVariableConstants.PluginPaths);
+            string[] paths;
+
+            if (string.IsNullOrEmpty(nugetPluginPaths))
+            {
+                nugetPluginPaths = _environmentVariableReader.GetEnvironmentVariable("PATH");
+            }
+
+            paths = nugetPluginPaths?.Split(Path.PathSeparator) ?? Array.Empty<string>();
+
+            foreach (var path in paths)
+            {
+                if (File.Exists(path))
+                {
+                    FileInfo fileInfo = new FileInfo(path);
+                    if (fileInfo.Name.StartsWith("nuget-plugin-", StringComparison.CurrentCultureIgnoreCase))
+                    {
+                        // A DotNet tool plugin
+                        if (IsValidPluginFile(fileInfo))
+                        {
+                            PluginFile pluginFile = new PluginFile(fileInfo.FullName, new Lazy<PluginFileState>(() => PluginFileState.Valid), isDotnetToolsPlugin: true);
+                            pluginFiles.Add(pluginFile);
+                        }
+                    }
+                    else
+                    {
+                        // A non DotNet tool plugin file
+                        var state = new Lazy<PluginFileState>(() => _verifier.IsValid(fileInfo.FullName) ? PluginFileState.Valid : PluginFileState.InvalidEmbeddedSignature);
+                        pluginFiles.Add(new PluginFile(fileInfo.FullName, state));
+                    }
+                }
+                else if (Directory.Exists(path))
+                {
+                    var directoryInfo = new DirectoryInfo(path);
+                    var files = directoryInfo.GetFiles("nuget-plugin-*");
```

Comment:
It looks like [`Directory.GetFiles`](https://learn.microsoft.com/dotnet/api/system.io.directory.getfiles?view=net-8.0) can throw several exceptions. I think the 2 important exceptions are [UnauthorizedAccessException](https://learn.microsoft.com/dotnet/api/system.unauthorizedaccessexception?view=net-8.0) and [PathTooLongException](https://learn.microsoft.com/dotnet/api/system.io.pathtoolongexception?view=net-8.0). These 2 exceptions are also caught explicitly in the doc that guides on [How to: Enumerate directories and files](https://learn.microsoft.com/en-us/dotnet/standard/io/how-to-enumerate-directories-and-files). In our case, we should at least log to notify users that a plugin has been ignored at least when [plugin diagnostic logging](https://github.com/NuGet/NuGet.Client/blob/dev/docs/plugin-logging.md) is enabled.

Reply:
> In our case, we should at least log to notify users that a plugin has been ignored due to an exception, at least when [plugin diagnostic logging](https://github.com/NuGet/NuGet.Client/blob/dev/docs/plugin-logging.md) is enabled.

I am not sure if a logger is passed to this API. If there is no access to the logger, I am okay if you would like to address this part of the feedback in a follow-up PR. It would be great if the feedback related to exceptions is handled in this PR.

Reply:
As you mentioned there is not loggers passed to PluginDiscoverer. This is also reflected in the current implementation : https://github.com/NuGet/NuGet.Client/blob/428d2d0013971187ac5f4ce4fc6402f191cf28e0/src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs#L84
This is a similar operation that could result in an exception being thrown. I think we should definitely introduce proper exception handling and logging in PluginDiscoverer. However, that could be done as a separate effort. 

We could use https://github.com/NuGet/NuGet.Client/blob/428d2d0013971187ac5f4ce4fc6402f191cf28e0/src/NuGet.Core/NuGet.Protocol/Plugins/Logging/PluginLogger.cs#L71 and log the exceptions. However, the purpose of that logger is to log plugin messages and updates using a file stream.

Reply:
After some digging I was able to find out that we do catch the exception at a higher level: https://github.com/NuGet/NuGet.Client/blob/428d2d0013971187ac5f4ce4fc6402f191cf28e0/src/NuGet.Core/NuGet.Protocol/HttpSource/HttpSourceAuthenticationHandler.cs#L291.
We also do logging on it. From my debug executions, this logging was not happening because the logger passed down here is a `NullLogger`.

Reply:
similar issue found: https://github.com/NuGet/Home/issues/12068